library/core/src/iter/adapters/intersperse.rs

```diff
@@ -33,13 +33,17 @@ where
     type Item = I::Item;
 
     #[inline]
-    fn next(&mut self) -> Option<I::Item> {
-        if self.needs_sep && self.iter.peek().is_some() {
-            self.needs_sep = false;
-            Some(self.separator.clone())
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(v) = self.next_item.take() {
+            Some(v)
         } else {
-            self.needs_sep = true;
-            self.iter.next()
+            let next_item = self.iter.next();
```

Comment:
for an empty iterator, this calls `next` after it's already returned `None` in the constructor, which I don't think you're supposed to do here without wrapping it in a `.fuse()` to make sure it doesn't return `Some` again after the `None`.

Reply:
Thanks @Nilstrieb, I wasn't even thinking about it. Could you check if my `fuse()` usage is as expected? Thx!

Reply:
I think that's good, but may make it slower. Could you attach some benchmarking results to the PR?

Reply:
@Nilstrieb https://github.com/nyurik/intersperse_perf

Reply:
P.S. I re-ran benchmarks a few times - the results are still about the same - my understanding is that `fuse()` becomes a noop if the underlying iterator guarantees proper None responses.

Reply:
(Note that it's fine if the Interperse wrapper isn't fused if the underlying iterator isn't, but then it is important that the inner .next() isn't called more often than the outer .next().)